app/models/episode.rb

```diff
@@ -133,28 +129,75 @@ def author=(a)
     self.author_email = author["email"]
   end
 
-  def enclosure
-    enclosures.complete.first
+  def ready_image
+    images.complete_or_replaced.first
   end
 
   def image
-    images.complete.first
-  end
-
-  # API updates for image=
-  def image_file
     images.first
   end
 
-  def image_file=(file)
+  def image=(file)
     img = EpisodeImage.build(file)
-    if img && img.original_url != image_file.try(:original_url)
-      images << img
-    elsif !img
-      images.destroy_all
+
+    if !img
+      images.each(&:mark_for_destruction)
+    elsif img&.replace?(image)
+      images.build(img.attributes.compact)
+    else
+      img.update_image(image)
+    end
+  end
+
+  def ready_enclosure
+    enclosures.complete_or_replaced.first
+  end
+
+  def enclosure
+    enclosures.first
+  end
+
+  def enclosure=(file)
+    enc = Enclosure.build(file)
+
+    if !enc
+      enclosures.each(&:mark_for_destruction)
+    elsif enc&.replace?(enclosure)
+      enclosures.build(enc.attributes.compact)
+    else
+      enc.update_resource(enclosure)
     end
   end
 
+  def ready_contents
+    contents.complete_or_replaced.group_by(&:position).values.map(&:first)
+  end
+
+  def contents=(files)
+    existing = contents.group_by(&:position)
```

Comment:
Currently, we do have multiple contents per position.  And we just grab the newest-completed one to say "this should be in the feed/API".  (We're a bit inconsistent even on that, though).

The `after_create` hook will start to sort that out, as episodes get re-saved (each Content will delete/replace its predecessors).

I think that will be okay?  This code also handles that case - serving up the newest content in that position.  That was my intent anyways.  Anything newly saved transitions to the new state, and old data continues to work as-is.

Was hoping to avoid "integer versioning" of files if possible.  That just feels like a bit overkill for us.  File-consistency only matters after an episode is published, and even then, users are generally only swapping out files.  If users are doing something more drastic, they should be unpublishing the episode while they fix it.

I may need to think a bit more about about handling old data...

Reply:
But the `has_many :contents` is effectively append-only, right?  We're just setting deleted_at/replaced_at timestamps on all but the 1st item (by position).

Reply:
Ahh ok,  I thought that guid was keyed on the episode.  I updated the code example.

One other thing to consider here is that if validations push against CMS too much, we could shim in validations that only log errors.